src/compiler/program.ts

```diff
@@ -1537,21 +1537,24 @@ namespace ts {
                     const emitFilePath = toPath(emitFileName, currentDirectory, getCanonicalFileName);
                     // Report error if the output overwrites input file
                     if (filesByName.contains(emitFilePath)) {
-                        let chain: DiagnosticMessageChain;
-                        if (!options.configFilePath) {
-                            // The program is from either an inferred project or an external project
-                            chain = chainDiagnosticMessages(/*details*/ undefined, Diagnostics.Adding_a_tsconfig_json_file_will_help_organize_projects_that_contain_both_TypeScript_and_JavaScript_files_Learn_more_at_https_Colon_Slash_Slashaka_ms_Slashtsconfig);
+                        if (options.noEmitOverwritenFiles && !options.out && !options.outDir && !options.outFile) {

```

Comment:
That is true for now, though the relationship might be loose. It is possible that in the future this option can be enabled for other projects, while this checking guarantees correctness without costing much